nibabel/funcs.py

```diff
@@ -122,8 +124,13 @@ def concat_images(images, check_affines=True):
         if check_affines:
             if not np.all(img.affine == affine):
                 raise ValueError('Affines do not match')
-        out_data[i] = img.get_data()
-    out_data = np.rollaxis(out_data, 0, len(i0shape)+1)
+        out_data[i] = img.get_data().copy()
+    if axis is not None:
+        out_data = np.concatenate(out_data, axis=axis)
```

Comment:
First - I think this makes another copy of the array, which may well be large by this point.

Second - isn't this the wrong 'axis'?   For example, the most common use here would be to concatenate a (i, j, k, N1) and an (i, j, k, N2) image.   That would be axis=-1 or axis=3.  I think the code above would first - fail if N1 != N2 in the `out_data[i] = ` line and if N1 == N2, would fail for axis=3 because the last allowed axis in the call above would be 2 (it is iterating over the first axis generating 3D arrays).   For axis=-1 (corresponding to axis == 2) it would generate something of shape (i, j, k * (N1 + N2)) which is still not what you want, I don't think.

I think the solution is to use concatenate without `out_data` if `axis != None` - so have separate code paths for the previous case and the new case.